File: src/main/java/io/transwarp/logparser/filter/DuplicationFilter.java
package io.transwarp.logparser.filter;

import io.transwarp.logparser.util.Record;

import java.util.HashSet;
import java.util.Set;

/**
 * Author: stk
 * Date: 18/3/31
 */
public class DuplicationFilter implements Filter {
    private Set<String> existed = new HashSet<>();

    @Override
    public boolean filter(Record record) {
        if (record.getDuplicationIdentifier() == null) return false;
        return existed.add(record.getDuplicationIdentifier());
    }

    @Override
    public void clean() {
        existed.clear();
    }
}

File: src/main/scala/io/transwarp/logparser/filter/DuplicationFilter.scala
package io.transwarp.logparser.filter

import io.transwarp.logparser.util.LogEntry

/**
  * Author: stk
  * Date: 2018/4/4
  */
class DuplicationFilter(override val params: List[String]) extends Filter {
  private var existed: Set[String] = Set()

  override def filter(logEntry: LogEntry): Boolean = logEntry.duplicationIdentifier
    .fold(true)(f => if (existed.contains(f)) false else {
      existed += f
      true
    })

  override def copy(): Filter = new DuplicationFilter(params)
}
